### ISSUE TITLE
splitter's split method should accept a list of string. Alternatively, allow splitting on DataFrames instead of dc.DataSet

### ISSUE THREAD

**Issue by MridulaBontha** [deepchem/deepchem#3221]
## 🚀 Feature

Allow DataSet or DataFrame object to be passed to split method.

## Motivation

Allowing splitting on DataFrames before featurization. In this case, user must provide 'smiles_col' as input.
Also, dc.featurizer allows multiple ids to be assigned for a datapoint. But dc splitters expect ids to be smiles.
This issue has already been raised in the [PR 1391](https://github.com/deepchem/deepchem/pull/1391).
Please add the feature in the master branch since PR-1391 was never merged before.

**Comment by rbharath** [comment]
@MridulaBontha Would you be interested in helping contribute the fix? #1391 got most of the way but it looks like tests need to be added still. If you don't have bandwidth, I'll mark this as an enhancement and put it on our docket and we will try to get to it as soon as we can.

**Comment by MridulaBontha** [comment]
I have a version of splitter that works on pandas dataframe instead of a deepchem dataset. I was successfully able to test for train_valid_test splits but not for k_fold splits. Can I coordinate with someone from your team to discuss the solution?

**Comment by gusty1g** [comment]
Two possible solutions: 
- As in #1391, add an additional parameter like `array` to the `split` method for accepting a array like object
- Overload the existing `dataset` attribute to accept a list of smiles string or an array or a pandas.Series object (anything which is array-like) and do type inference inside the `split` method to provide the desired functionality.

I prefer the later approach since in the former one, having two arguments for dataset - a `array` and a `dataset` will cause ambiguity to users.

Other notes:
- Splitter exposes three common methods for end users: `train_test_split`, `train_test_valid_split`, `k_fold_split` which internally calls `split` method. Some splitters like `StratifiedSplitter` depends on y values. How do we pass y values to these splitters when we are splitting an array like object? Maybe we can use kwargs or as an additional argument for these splitters.

**Comment by MridulaBontha** [comment]
I agree with you, I have written some code following approach 2. I am using pandas df as input along with smiles_col and label_col as intput. Expectation is that all these values should exist in pandas df only. 
I am having trouble putting code for kfold splits. Shall we work together to fix this? I can share code with you!

**Comment by gusty1g** [comment]
Feel free to land a PR with your code and we can continue from there.
